File: ts/files/src/core/exceptions.ts
/**
 * Custom error classes for Figma Files API
 * Provides structured error handling with context and metadata
 */

/**
 * Base API error class for all Figma API related errors
 */
export class FigmaApiError extends Error {
  code: string;
  meta: Record<string, any>;
  timestamp: string;

  constructor(message: string, code: string, meta: Record<string, any> = {}) {
    super(message);
    this.name = this.constructor.name;
    this.code = code;
    this.meta = meta;
    this.timestamp = new Date().toISOString();
  }
}

/**
 * Rate limit exceeded error
 * Thrown when API rate limits are hit
 */
export class RateLimitError extends FigmaApiError {
  retryAfter: number | null;

  constructor(
    retryAfter: number | null = null,
    requestId: string | null = null,
  ) {
    const message = retryAfter
      ? `Rate limit exceeded. Retry after ${retryAfter} seconds`
      : "Rate limit exceeded";

    super(message, "RATE_LIMIT_EXCEEDED", {
      retryAfter,
      requestId,
      retryable: true,
    });
    this.retryAfter = retryAfter;
  }
}

/**
 * Authentication error
 * Thrown when API token is invalid or missing
 */
export class AuthenticationError extends FigmaApiError {
  constructor(message: string = "Authentication failed") {
    super(message, "AUTHENTICATION_FAILED", { retryable: false });
  }
}

/**
 * Authorization error
 * Thrown when user doesn't have permission to access resource
 */
export class AuthorizationError extends FigmaApiError {
  constructor(
    message: string = "Insufficient permissions",
    requiredScopes: any[] = [],
  ) {
    super(message, "AUTHORIZATION_FAILED", {
      requiredScopes,
      retryable: false,
    });
  }
}

/**
 * File not found error
 * Thrown when file_key doesn't exist or isn't accessible
 */
export class FileNotFoundError extends FigmaApiError {
  constructor(fileKey: string) {
    super(`File not found: ${fileKey}`, "FILE_NOT_FOUND", {
      fileKey,
      retryable: false,
    });
  }
}

/**
 * Node not found error
 * Thrown when node IDs don't exist in the file
 */
export class NodeNotFoundError extends FigmaApiError {
  constructor(nodeIds: string | string[], fileKey: string) {
    const ids = Array.isArray(nodeIds) ? nodeIds.join(", ") : nodeIds;
    super(`Nodes not found: ${ids}`, "NODES_NOT_FOUND", {
      nodeIds,
      fileKey,
      retryable: false,
    });
  }
}

/**
 * Validation error
 * Thrown when request parameters are invalid
 */
export class ValidationError extends FigmaApiError {
  constructor(message: string, field: string | null = null, value: any = null) {
    super(message, "VALIDATION_ERROR", {
      field,
      value,
      retryable: false,
    });
  }
}

/**
 * Network error
 * Thrown when network requests fail
 */
export class NetworkError extends FigmaApiError {
  originalError: Error | null;

  constructor(message: string, originalError: Error | null = null) {
    super(message, "NETWORK_ERROR", {
      originalError: originalError?.message,
      retryable: true,
    });
    this.originalError = originalError;
  }
}

/**
 * HTTP error
 * Thrown for various HTTP status codes
 */
export class HttpError extends FigmaApiError {
  status: number;
  statusText: string;
  url: string;

  constructor(
    status: number,
    statusText: string,
    url: string,
    body: any = null,
  ) {
    super(`HTTP ${status}: ${statusText}`, "HTTP_ERROR", {
      status,
      statusText,
      url,
      body,
      retryable: status >= 500 || status === 429,
    });
    this.status = status;
    this.statusText = statusText;
    this.url = url;
  }
}

/**
 * Server error
 * Thrown for 5xx status codes
 */
export class ServerError extends FigmaApiError {
  constructor(
    message: string = "Internal server error",
    requestId: string | null = null,
  ) {
    super(message, "SERVER_ERROR", {
      requestId,
      retryable: true,
    });
  }
}

/**
 * Timeout error
 * Thrown when requests exceed timeout threshold
 */
export class TimeoutError extends FigmaApiError {
  constructor(timeout: number) {
    super(`Request timed out after ${timeout}ms`, "TIMEOUT", {
      timeout,
      retryable: true,
    });
  }
}

/**
 * Utility function to create appropriate error from HTTP response
 * @param {Response} response - Fetch API response
 * @param {string} url - Request URL
 * @param {Object} body - Parsed response body
 * @returns {FigmaApiError} Appropriate error instance
 */
export function createErrorFromResponse(
  response: any,
  url: string,
  body: any = null,
): FigmaApiError {
  const { status, statusText } = response;
  const requestId = response.headers.get("x-request-id");

  switch (status) {
    case 401:
      return new AuthenticationError(body?.message || "Invalid API token");

    case 403:
      return new AuthorizationError(
        body?.message || "Insufficient permissions",
        body?.requiredScopes,
      );

    case 404: {
      // Try to determine if it's a file or node error from URL
      const fileKeyMatch = url.match(/\/files\/([^/]+)/);
      if (fileKeyMatch) {
        return new FileNotFoundError(fileKeyMatch[1]);
      }
      return new FigmaApiError("Resource not found", "NOT_FOUND", {
        status,
        url,
      });
    }

    case 429: {
      const retryAfter = response.headers.get("retry-after");
      return new RateLimitError(
        retryAfter ? parseInt(retryAfter, 10) : null,
        requestId,
      );
    }

    case 400:
      return new ValidationError(body?.message || "Bad request", null, body);

    case 500:
    case 502:
    case 503:
    case 504:
      return new ServerError(body?.message || "Server error", requestId);

    default:
      return new HttpError(status, statusText, url, body);
  }
}

/**
 * Check if an error is retryable
 * @param {Error} error - Error to check
 * @returns {boolean} Whether the error is retryable
 */
export function isRetryableError(error: any): boolean {
  if (error instanceof FigmaApiError) {
    return error.meta.retryable || false;
  }

  // Network errors are generally retryable
  if (error instanceof NetworkError || error.code === "NETWORK_ERROR") {
    return true;
  }

  return false;
}

export default {
  FigmaApiError,
  RateLimitError,
  AuthenticationError,
  AuthorizationError,
  FileNotFoundError,
  NodeNotFoundError,
  ValidationError,
  NetworkError,
  HttpError,
  ServerError,
  TimeoutError,
  createErrorFromResponse,
  isRetryableError,
};

File: packages/py/comments/src/figma_comments/core/exceptions.py
"""Custom exceptions for Figma Comments API operations."""

from typing import Optional, Dict, Any


class FigmaCommentsError(Exception):
    """Base exception for all Figma Comments API errors."""

    def __init__(
        self, 
        message: str, 
        status_code: Optional[int] = None,
        response_data: Optional[Dict[str, Any]] = None
    ) -> None:
        super().__init__(message)
        self.message = message
        self.status_code = status_code
        self.response_data = response_data or {}

    def __str__(self) -> str:
        if self.status_code:
            return f"[{self.status_code}] {self.message}"
        return self.message


class AuthenticationError(FigmaCommentsError):
    """Raised when authentication fails (401 Unauthorized)."""

    def __init__(
        self, 
        message: str = "Invalid or missing API token",
        response_data: Optional[Dict[str, Any]] = None
    ) -> None:
        super().__init__(message, 401, response_data)


class AuthorizationError(FigmaCommentsError):
    """Raised when authorization fails (403 Forbidden)."""

    def __init__(
        self, 
        message: str = "Insufficient permissions for this operation",
        response_data: Optional[Dict[str, Any]] = None
    ) -> None:
        super().__init__(message, 403, response_data)


class NotFoundError(FigmaCommentsError):
    """Raised when a resource is not found (404 Not Found)."""

    def __init__(
        self, 
        message: str = "Requested resource not found",
        response_data: Optional[Dict[str, Any]] = None
    ) -> None:
        super().__init__(message, 404, response_data)


class RateLimitError(FigmaCommentsError):
    """Raised when rate limit is exceeded (429 Too Many Requests)."""

    def __init__(
        self, 
        message: str = "Rate limit exceeded",
        retry_after: Optional[int] = None,
        response_data: Optional[Dict[str, Any]] = None
    ) -> None:
        super().__init__(message, 429, response_data)
        self.retry_after = retry_after

    def __str__(self) -> str:
        base_msg = super().__str__()
        if self.retry_after:
            return f"{base_msg} (retry after {self.retry_after} seconds)"
        return base_msg


class ValidationError(FigmaCommentsError):
    """Raised when request validation fails (400 Bad Request)."""

    def __init__(
        self, 
        message: str = "Request validation failed",
        errors: Optional[list] = None,
        response_data: Optional[Dict[str, Any]] = None
    ) -> None:
        super().__init__(message, 400, response_data)
        self.errors = errors or []

    def __str__(self) -> str:
        base_msg = super().__str__()
        if self.errors:
            error_details = "; ".join(str(error) for error in self.errors)
            return f"{base_msg}: {error_details}"
        return base_msg


class ApiError(FigmaCommentsError):
    """Generic API error for unexpected status codes."""

    def __init__(
        self, 
        message: str,
        status_code: int,
        response_data: Optional[Dict[str, Any]] = None
    ) -> None:
        super().__init__(message, status_code, response_data)


class NetworkError(FigmaCommentsError):
    """Raised when network/connection errors occur."""

    def __init__(
        self, 
        message: str = "Network error occurred",
        original_exception: Optional[Exception] = None
    ) -> None:
        super().__init__(message)
        self.original_exception = original_exception

    def __str__(self) -> str:
        base_msg = super().__str__()
        if self.original_exception:
            return f"{base_msg}: {self.original_exception}"
        return base_msg


class TimeoutError(FigmaCommentsError):
    """Raised when request timeout occurs."""

    def __init__(
        self, 
        message: str = "Request timeout",
        timeout_seconds: Optional[float] = None
    ) -> None:
        super().__init__(message)
        self.timeout_seconds = timeout_seconds

    def __str__(self) -> str:
        base_msg = super().__str__()
        if self.timeout_seconds:
            return f"{base_msg} ({self.timeout_seconds}s)"
        return base_msg


class CommentNotFoundError(NotFoundError):
    """Raised when a specific comment is not found."""

    def __init__(
        self, 
        comment_id: str,
        file_key: str,
        response_data: Optional[Dict[str, Any]] = None
    ) -> None:
        message = f"Comment '{comment_id}' not found in file '{file_key}'"
        super().__init__(message, response_data)
        self.comment_id = comment_id
        self.file_key = file_key


class FileNotFoundError(NotFoundError):
    """Raised when a Figma file is not found."""

    def __init__(
        self, 
        file_key: str,
        response_data: Optional[Dict[str, Any]] = None
    ) -> None:
        message = f"Figma file '{file_key}' not found or not accessible"
        super().__init__(message, response_data)
        self.file_key = file_key


class InvalidCommentError(ValidationError):
    """Raised when comment data is invalid."""

    def __init__(
        self, 
        message: str = "Invalid comment data",
        field_errors: Optional[Dict[str, str]] = None,
        response_data: Optional[Dict[str, Any]] = None
    ) -> None:
        super().__init__(message, response_data=response_data)
        self.field_errors = field_errors or {}

    def __str__(self) -> str:
        base_msg = super().__str__()
        if self.field_errors:
            field_details = "; ".join(f"{field}: {error}" for field, error in self.field_errors.items())
            return f"{base_msg}: {field_details}"
        return base_msg